### ISSUE TITLE
OnDisplaying in custom ShapeTableProvider does not populate data anymore in some cases

### ISSUE THREAD

**Issue by Habbni** [OrchardCMS/OrchardCore#16673]
### Describe the bug

after upgrading from v1.8.3 to the current build, we can no longer use a custom ShapeTableProvider to add missing alternates for the shape ContentsAdminListCreate using following pattern:

```
builder.Describe("ContentsAdminListCreate")
    .OnDisplaying(async context =>
    {
        var differentiator = (context.Shape as ContentOptionsViewModel)?.SelectedContentType?.EncodeAlternateElement();
        if (differentiator != null)
        {
            context.Shape.Metadata.Alternates.Add(differentiator + "__ContentsAdminListCreate");
        }
    });
```

This worked perfectly and it still does for almost any other shape, so we suspect this to be unintended.
With the current preview-2.0.0, or running latest main, context.Shape does not contain any of the shape/model data.
In this case, for example CreatableTypes and SelectedContentType remain empty.

It seems to only affect those shapes created using Initialize, like in this case:

`Initialize<ContentOptionsViewModel>("ContentsAdminListCreate", m =>  `

Shapes created without Initialize still get their properties populated, for example this one:

`results.Add(Shape("ContentsButtonActions_SummaryAdmin", new ContentItemViewModel(contentItem)).Location("SummaryAdmin", "ActionsMenu:10")`

### Orchard Core version

2.0.0-preview-18331

### To Reproduce
1. Take the latest build (it worked only a few builds ago)
2. Create a custom ShapeTableProvider and register it on Startup
3. Override DiscoverAsync, add above Describe method
4. Set a breakpoint within OnDisplaying, go to any List, e.g. the ContentItems and see no shape related data is being filled in

### Expected behavior
We would expect to see context.Shape properties to be populated, like it used to do, and like it still does when adding (for example) the  following to the ShapeTableProvider:
```
builder.Describe("ContentsButtonActions_SummaryAdmin")
.OnDisplaying(async context =>{}
```

**Comment by MikeAlhayek** [comment]
@Habbni, are you by any chance using caching for your shapes?

We encountered an issue where the initializer (e.g., `Initialize<ContentOptionsViewModel>("ContentsAdminListCreate", m =>`) was being called every time, even when the object was cached. This was resolved in PR #16644, which ensures the initializer is only called if the shape is not cached.

I believe this change in behavior might be what you're experiencing.

**Comment by MikeAlhayek** [comment]
@sebastienros I have not tried to reproduce the issue. But just point out I don't see us using cache:

https://github.com/OrchardCMS/OrchardCore/blob/4f5cd18ba6bdb10f95f8af6a584035c3f9085eb1/src/OrchardCore.Modules/OrchardCore.Contents/Drivers/ContentOptionsDisplayDriver.cs#L35

**Comment by Habbni** [comment]
Hi and thanks for your reply

No caching configured. I actually downloaded a new clone and did a fresh blog recipe setup, to make sure it is reproducable. Just tried disabling the cache module because your reply sounds reasonable, but the issue persists..

![grafik](https://github.com/user-attachments/assets/c2afa683-f62d-4639-9d00-11acc4fae95e)

**Comment by MikeAlhayek** [comment]
@sebastienros can you please take a look at this during triage please?

**Comment by sebastienros** [comment]
Will check. @Habbni can you try at commits 
1- 8fd5c885d2b713da2a9c982e70297b12b5027c6a
2- b316211456ec4eb251576f8625a6079fef83f4fb

to confirm this is the change we suspect?

**Comment by Habbni** [comment]
**works** with 8fd5c88:
![grafik](https://github.com/user-attachments/assets/0b3f4532-0442-4225-a06b-5b7f2938228c)

**doesnt** with b316211 (see screenshot above)

thanks a lot already guys, really wish i was deeper into that code and could contribute more help than this..

**Comment by sebastienros** [comment]
I will debug to understand what is failing, or if the usage described in this issue is that actual problem and another event should have been used.

**Comment by sebastienros** [comment]
Thanks @Habbni for confirming that quickly, very helpful.

**Comment by Habbni** [comment]
> I will debug to understand what is failing, or if the usage described in this issue is that actual problem and another event should have been used.

I tried the other events as well. But since it was before posting this issue and my brain reached it´s 2bit capacity, I´m not sure if i remember correctly: only OnDisplayed came up with data, which seemed too late for adding an alternate..

**Comment by sebastienros** [comment]
Correct.

0. Displaying
1. Processing
2. Displayed

I moved the initialization in Processing on the commit you pointed out. I probably should have put it in Displaying since it should not be called on Cache either. Then it will only be a matter of which "Displaying" is called first.

**Comment by sebastienros** [comment]
- Dynamic Cache uses the `DisplayingAsync` event to populate the cache, hence these events need to be invoked all the time
- ProcessingAsync was made specifically for code that needs to be executed unless it's cached. Because of that they are executed after `DisplayAsync`. If you look at `MenuShapes` it uses `ProcessAsync` and not `DisplayAsync` to create the menu nodes specifically for this reason.

What I did for now is invoking the code from Initialize as the first thing done in `DisplayAsync` to emulate what was done before my PR. I will also move the code from the Notification that is cache to be in ProcessingAsync because we want to run it only when not cached. It may require a new helper on drivers, but I had already suggested it.
